### DISCUSSION TITLE
ADB in WSL2

### DISCUSSION THREAD

**Discussion by yanotoma** [microsoft/WSL#4692] [General]
ADB (Android Debug Bridge) used to work on WSL1 but no longer in WSL2
The catch was to use the same version in both windows and linux

**WSL1**
![image](https://user-images.githubusercontent.com/2429644/69091232-dd353a00-0a17-11ea-8123-595ad9f52478.png)

**WSL2**
![image](https://user-images.githubusercontent.com/2429644/69092413-52097380-0a1a-11ea-9a3c-9dff22d2144a.png)

**OS**
![image](https://user-images.githubusercontent.com/2429644/69091472-5af94580-0a18-11ea-8bad-8ab3d6f16bb5.png)

**Comment by ghost** [comment]
Hi! We've identified this issue as a duplicate of another one that already exists in this repository. This specific instance is being closed in favor of tracking the concern over on the referenced thread.

Thanks for your report!

**Comment by therealkenc** [comment]
TCP bridge. Likely dupe #4619 but having pulled the trigger too fast once already will let it float.

**Comment by andreialecu** [comment]
See https://github.com/microsoft/WSL/issues/4619#issuecomment-596990676 for a workaround.

**Reply by m4v3r** [reply]
this worked with no issues, didn't even check any versions... ;)

**Comment by gusflopes** [comment]
Workaround didn't work for me...

**Comment by anaganisk** [comment]
@gusflopes For me it was windows firewall issue, check  your firewall settings, after correcting that it worked.

**Comment by steelbrain** [comment]
~~Fwiw, what worked for me is to start the ADB server in WSL and make the Host Windows connect to it. Set `ADB_SERVER_SOCKET=tcp:localhost:5037` in your windows environment variables.~~

~~Also make sure that your host and WSL adb are the same version or else they'll keep killing each other's servers~~

Alright so here's what you're gonna do:

Add the following to your WSL bashrc or zshrc:

```
export WSL_HOST_IP="$(tail -1 /etc/resolv.conf | cut -d' ' -f2)"
export ADB_SERVER_SOCKET=tcp:$WSL_HOST_IP:5037
```

Then create a firewall entry, create a new "Inbound" Rule. Select "Port" then Specific TCP port "5037" then "Allow the connection" then check all of Domain, Private and Public and add a name. After the firewall entry is added, open up its properties, go to Scope -> Remote IP Addresses -> Add "172.16.0.0/12".

Now that we're covered in the WSL2 VM and through the Firewall, you have to start the adb server with specific arguments to make it listen on all addresses (don't worry we've only whitelisted our WSL2 VM IP range so no security issues).

Create a vbs script, call it whatever and put this in it
```
CreateObject("WScript.Shell").Run "%USERPROFILE%\AppData\Local\Android\Sdk\platform-tools\adb.exe kill-server", 0, True
CreateObject("WScript.Shell").Run "%USERPROFILE%\AppData\Local\Android\Sdk\platform-tools\adb.exe -a -P 5037 nodaemon server", 0, True
```

Now all you have to do is invoke the vbs script once per reboot and your WSL VM will connect to your host ADB instance

**Reply by danielkv** [reply]
I did it all, but in the end: "Couldn't start project on Android: could not connect to TCP port 5556: Connection refused"

**Reply by warifp** [reply]
> 172.16.0.0/12

Thank!

**Reply by Monk3y51** [reply]
> ~Fwiw, what worked for me is to start the ADB server in WSL and make the Host Windows connect to it. Set `ADB_SERVER_SOCKET=tcp:localhost:5037` in your windows environment variables.~
> 
> ~Also make sure that your host and WSL adb are the same version or else they'll keep killing each other's servers~
> 
> Alright so here's what you're gonna do:
> 
> Add the following to your WSL bashrc or zshrc:
> 
> ```
> export WSL_HOST_IP="$(tail -1 /etc/resolv.conf | cut -d' ' -f2)"
> export ADB_SERVER_SOCKET=tcp:$WSL_HOST_IP:5037
> ```
> 
> Then create a firewall entry, create a new "Inbound" Rule. Select "Port" then Specific TCP port "5037" then "Allow the connection" then check all of Domain, Private and Public and add a name. After the firewall entry is added, open up its properties, go to Scope -> Remote IP Addresses -> Add "172.16.0.0/12".
> 
> Now that we're covered in the WSL2 VM and through the Firewall, you have to start the adb server with specific arguments to make it listen on all addresses (don't worry we've only whitelisted our WSL2 VM IP range so no security issues).
> 
> Create a vbs script, call it whatever and put this in it
> 
> ```
> CreateObject("WScript.Shell").Run "%USERPROFILE%\AppData\Local\Android\Sdk\platform-tools\adb.exe kill-server", 0, True
> CreateObject("WScript.Shell").Run "%USERPROFILE%\AppData\Local\Android\Sdk\platform-tools\adb.exe -a -P 5037 nodaemon server", 0, True
> ```
> 
> Now all you have to do is invoke the vbs script once per reboot and your WSL VM will connect to your host ADB instance

This worked perfectly for me but to save me from having to run the script each time i added to script to a schedule to run at logon

I opened Task Scheduler

Create Basic Task

Add name and description (helpful to find it later if you want to delete it)

Set the trigger to "When I log on"

Leave Action as "Start a program"

Select the VBS script you created, (i placed mine in a folder named scripts that was out the way so i did not accidentally delete it)

Click finish and thats it the script will run when you login ready for if you use ADB on WSL


NOTE: When i created the VBS script it did not work for me as the adb location referenced in it was different to mine so changed the path to where ADB is located

Thank you for this solution i have been searching for weeks at different ways to get ADB working on WSL

**Comment by Newbie012** [comment]
I was able to connect over wifi without the need of installing on both sides. See - https://developer.android.com/studio/command-line/adb/?gclid=CjwKCAjw2dD7BRASEiwAWCtCbxKzwu63T-HpLaIp8ASO1aNA6aRl7_8Wvc2LqCvf3BI3umOlQQaOtBoCQrEQAvD_BwE&gclsrc=aw.ds#wireless

**Comment by therealkenc** [comment]
This one ("used to work on WSL1 but no longer in WSL2") is #4619. It's analogous X11 Server on Windows, Database server on Windows, etc etc. Rather than dupe, porting over to the discussion forum for tips/tricks ADB over TCP with WSL2. Thanks for the input everyone.

**Comment by skamensky** [comment]
I was using Intellij''s Android developement setup inside of WSL2. IntelliJ/Android Studio does not let you set the 
`ADB_SERVER_SOCKET` environment variable. It's hardcoded to `127.0.0.1:5037`. So the https://github.com/microsoft/WSL/discussions/4692#discussioncomment-87987 answer didn't work exactly for me.

The fix I came up with was to use `socat` to forward the windows 5037 port to the linux 5037 port.
```bash
export WSL_HOST_IP="$(tail -1 /etc/resolv.conf | cut -d' ' -f2)"
socat TCP-LISTEN:5037,reuseaddr,fork TCP:$WSL_HOST_IP:5037
```
Make sure to run this in a separate process *before* starting up Intellij/Android Studio, since the IDE will start up an adb server if it detects that one is not yet running. Once it starts up the server, you will not be able to run the above socat command, and it will result in an error of something to the effect of `socat[276040] E bind(5, {AF=2 0.0.0.0:5037}, 16): Address already in use`

**Reply by m4v3r** [reply]
I actually updated WSL and Ubuntu and basically reset the environment to defaults and now everything just works between Linux and Windows network -side as well as USB passthrough with no futher configuration or anything. (different project than yours but I had similar problems)

**Reply by ninh-huynh** [reply]
The `socat` command works. Now I'm able to run Android Studio inside WSL2, and connect the emulator from Windows host. Without `socat`, we are only able to use `adb` in WSL2.

**Comment by eynopv** [comment]
Getting `error: could not connect to TCP port 5554: Connection refused` on `adb emu`.
The `adb devices` shows list of available devices and I am able to use `adb shell`.

I have set those environment variables in WSL
```
export WSL_HOST=$(tail -1 /etc/resolv.conf | cut -d' ' -f2)
export ADB_SERVER_SOCKET=tcp:$WSL_HOST:5037
```

And starting `adb` in windows as
```
adb -a -P 5037 nodaemon server start
```

**WSL2 distribution**
Linux eynopv-pc 5.15.90.1-microsoft-standard-WSL2 #1 SMP Fri Jan 27 02:56:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

**adb Windows**
Android Debug Bridge version 1.0.41
Version 34.0.1-9979309

**adb WSL**
Android Debug Bridge version 1.0.41
Version 34.0.1-9979309

**Reply by jeanfortheweb** [reply]
same problem here, occured after upgrading react-native. did you find a solution?

**Reply by eynopv** [reply]
no

**Reply by zigang93** [reply]
same issue...

**Comment by bouncydragon** [comment]
After following some guide, I still cannot connect my WSL2 React Native build to the emulator of my Android Studio in Windows. The run returns me this message:
![image](https://github.com/microsoft/WSL/assets/9712523/3b077c7b-dc99-4a8d-a87e-310671d19231)

While the `adb -a nodaemon server start` returns me this:
![image](https://github.com/microsoft/WSL/assets/9712523/010a3ef7-de1d-4d26-987e-41ff75de8a2b)

**Comment by JhnBer** [comment]
After two days of research, here is what helped solve my problem in my case.
I used Netish interface portproxy to add a port proxy that connects the host port to the WSL virtual machine.
`netsh interface portproxy add v4tov4 listenport=<yourPortToForward> listenaddress=0.0.0.0 connectport=<yourPortToConnectToInWSL> connectaddress=(wsl hostname -I)`

In my case, the device used port 21503. Therefore, the command looked like:
`netsh interface portproxy add v4tov4 listenport=21503 listenaddress=0.0.0.0 connectport=21503 connectaddress=127.0.1.1`

Then run `cat /etc/resolv.conf` in WSL and use nameserver value with adb connect.

More information here https://learn.microsoft.com/en-us/windows/wsl/networking

In fact, for me this was the only option since I couldn't use WSL 1 (which connects to localhost without problems since it is not a VM).

**Reply by Cryxto** [reply]
should i use wsl2 mirrored network or just default network mode of wsl2 ?

**Comment by aleixrodriala** [comment]
For the wireless-device case mentioned a few times above: the setup that's been painless for me is to not run a Linux adb server inside WSL2 at all — call the Windows `adb.exe` through interop (e.g. `alias adb='/mnt/c/Users/<you>/AppData/Local/Android/Sdk/platform-tools/adb.exe'`) so Windows owns the single adb server, and connect the phone over Wi-Fi instead of fighting USB passthrough.

The remaining friction is `adb pair` — reading codes and ip:ports off the phone screen — and the existing QR-pairing tools don't work from WSL2 because they run their own mDNS listener, and LAN multicast never crosses the NAT. I built a small CLI that does Android-Studio-style QR pairing but delegates discovery to the adb server itself (`adb mdns services`), so it works fine from WSL2 driving `adb.exe`: https://github.com/aleixrodriala/adb-qr — `uvx adb-qr`, scan, paired and connected (disclosure: I'm the author).

No usbipd, no firewall rules, no port proxying.

**Comment by foxvirtual** [comment]
To fix all of those problems , try these.

`unset ADB_SERVER_SOCKET`

`ls "/mnt/c/Users/<<USER>>/AppData/Local/Android/Sdk/platform-tools/"`

You should see:

adb.exe
fastboot.exe


Then run `"/mnt/c/Users/<<USER>>/AppData/Local/Android/Sdk/platform-tools/adb.exe" devices`

You should get:

List of devices attached
emulator    device

Make it convenient

`alias adb='/mnt/c/Users/<<USER>>/AppData/Local/Android/Sdk/platform-tools/adb.exe'`
then run
adb devices

To make the alias permanent:

`echo "alias adb='/mnt/c/Users/<<USER>>/AppData/Local/Android/Sdk/platform-tools/adb.exe'" >> ~/.bashrc`
`source ~/.bashrc`
